scripts/merge_source_cat.py

```diff
@@ -0,0 +1,415 @@
+
+    mag, mag_err = calib.getMagnitude(cat[flux_names['psf_flux']].values,
+                                      cat[flux_names['psf_flux_err']].values)
+
+    cat['mag'] = mag
+    cat['mag_err'] = mag_err
+    cat['SNR'] = np.abs(cat[flux_names['psf_flux']] /
+                        cat[flux_names['psf_flux_err']])
+
+    # TODO Need to calibrate fluxes as well
+    # For now we'll just save this information in an additional column
+    # to be available further downstream.
+    flux_mag0, flux_mag0_err = calib.getFluxMag0()
+    cat['fluxmag0'] = flux_mag0
+
+    # Associate with closest
+    object_id = associate_object_ids(cat,
+                                     data_ref=data_ref,
+                                     object_table=object_table,
+                                     matching_radius=matching_radius,
+                                     **kwargs)
+    cat['objectId'] = object_id
+
+    # Restrict to columns that we need
+    cat = cat[columns_to_keep]
+
+    return cat
+
+
+def associate_object_ids(cat, data_ref=None, object_table=None, **kwargs):
+    """Wrapper for development to easily switch
+    Object-Table based
+    coadd file based
+
+    associate_object_ids
+    """
+    if object_table is None:
+        associated_ids = associate_object_ids_to_coadd(cat, data_ref=data_ref, **kwargs)
+    else:
+        associated_ids = associate_object_ids_to_table(cat, object_table=object_table, **kwargs)
+
+    return associated_ids
+
+
+def associate_object_ids_to_coadd(cat, data_ref=None, verbose=True, **kwargs):
+    """Load and match to deepcoadd references."""
+
+    # Can I get this directly from the data_ref, or do I need data_ref.butler?
+    skymap = data_ref.get(datasetType='deepCoadd_skyMap')
+
+    # radians
+    min_ra, max_ra = np.min(cat['coord_ra']), np.max(cat['coord_ra'])
+    min_dec, max_dec = np.min(cat['coord_dec']), np.max(cat['coord_dec'])
+
+    coord_list = [
+        SpherePoint(min_ra * radians, min_dec * radians),
+        SpherePoint(max_ra * radians, min_dec * radians),
+        SpherePoint(max_ra * radians, max_dec * radians),
+        SpherePoint(min_ra * radians, max_dec * radians),
+    ]
+
+    # This will be a list of tracts that overlap
+    # Each tract entry will have its own list of patches that overlap
+    tract_patches = skymap.findTractPatchList(coord_list)
+
+    associated_id_table = []
+    for tract, patches in tract_patches:
+        for patch in patches:
+            tract_patch_data_id = {}
+            tract_patch_data_id['tract'] = tract.getId()
+            # Yes, really, the following is the supported way to get the
+            # patch ID in a way that can be queried.
+            patch_str = '%d,%d' % patch.getIndex()
+            tract_patch_data_id['patch'] = patch_str
+            if verbose:
+                print("Searching ", tract_patch_data_id)
+            try:
+                ref_table = data_ref.getButler().get(datasetType='deepCoadd_ref',
```

Comment:
I see, you have to retrieve the butler object because you don't have the tract information until later. I wonder if there's something like the tracts mapping table (overlaps table) in `/global/projecta/projectdirs/lsst/global/in2p3/Run1.2i/w_2018_39/rerun/281118/tracts_mapping.sqlite3`  that allows you to select tracts and patches in advance.

Reply:
Not provided in the current DM Science Pipelines Gen2 Butler.